File: lambda/login/module/auth/confirm_code_forgot_password.go
package auth

import (
	"daita_module_login/component/common"
	usermodel "daita_module_login/dynamoDB/user/model"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"regexp"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
)

func (app Application) confirmCodeForgotPassword(username, password, confirmCode string) (bool, error) {
	// confirm code
	resq := cognitoidentityprovider.ConfirmForgotPasswordInput{
		Username:         aws.String(username),
		Password:         aws.String(password),
		ConfirmationCode: aws.String(confirmCode),
		ClientId:         aws.String(app.Configure.ClientPooID),
	}
	req, _ := app.Coginto.ConfirmForgotPasswordRequest(&resq)
	err := req.Send()

	//error
	if err != nil {
		fmt.Fprintf(os.Stderr, fmt.Sprintf("Error confirmation code for password recovery: %s", err.Error()))
		return false, err
	}

	return true, nil
}

func (app Application) sendConfirmCodeforgotHandler(event events.APIGatewayProxyRequest, headers map[string]string) (string, int, map[string]string, map[string]interface{}, error) {
	data := map[string]interface{}{}
	user := usermodel.UserSendConfirmForgotPassword{}
	err := json.Unmarshal([]byte(event.Body), &user)
	message := ""
	if err != nil {
		message = common.MessageUnmarshalInputJson
		return message, http.StatusOK, headers, data, common.ErrUnmarshalInputJson
	}
	// check password strength
	passwd := user.Password
	secure := true
	tests := []string{".{8,}", "[a-z]", "[A-Z]", "[0-9]", "[^\\d\\w]"}
	for _, test := range tests {
		t, _ := regexp.MatchString(test, passwd)
		if !t {
			secure = false
			break
		}
	}
	if !secure {
		message := common.MessageInvalidPassword
		return message, http.StatusOK, headers, data, common.ErrInvalidPassword
	}
	// check confirm code
	isSuccess, err := app.confirmCodeForgotPassword(user.UserName, user.Password, user.ConfirmCode)
	if err != nil {
		errString := fmt.Sprintf(err.Error())
		for index, pattern := range common.ErrCogintoResponse {
			res, _ := regexp.MatchString(pattern, errString)
			if res {
				message := fmt.Sprintf(common.ErrorMessageCoginto[index].Error())
				return message, http.StatusOK, headers, data, common.ErrorMessageCoginto[index]
			}
		}
		message = common.MessageForgotPasswordConfirmcodeFailed
		return message, http.StatusOK, headers, data, common.ErrSendEmailFailed
	}
	if isSuccess {
		message = common.MessageForgotPasswordConfirmcodeSuccessfully
		return message, http.StatusOK, headers, data, nil
	}
	message = common.MessageForgotPasswordConfirmcodeFailed
	return message, http.StatusOK, headers, data, common.ErrSendEmailFailed
}
func (app Application) ImpConfirmCodeForgotPassword(event events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {

	headers := map[string]string{"Access-Control-Allow-Credentials": "true",
		"Access-Control-Allow-Methods": "GET,HEAD,OPTIONS,POST,PUT",
	}
	message := ""
	statusCode := http.StatusOK
	err := common.ErrInit
	data := map[string]interface{}{}
	message, statusCode, headers, data, err = app.sendConfirmCodeforgotHandler(event, headers)
	return common.GenerateResponseBody(message, statusCode, headers, data, err, []string{}), nil
}

File: core_service/aws_lambda/project/code/confirm_code_forgot_password.py
import re

import boto3
import json
from config import *
from error import *
from custom_mail import *
from response import generate_response, error_response


cog_provider_client = boto3.client("cognito-idp")
RESPONSE_HEADER = {
    "Access-Control-Allow-Credentials": "true",
    "Access-Control-Allow-Methods": "GET, HEAD, OPTIONS, POST, PUT",
}
PASSWORD_REGEX = r"^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[\^$*.\[\]{}\(\)?\-\"!@#%&\/,><\':;|_~`])\S{8,99}$"


@error_response
def lambda_handler(event, context):
    """
    https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-concepts.html#gettingstarted-concepts-event
    https://docs.aws.amazon.com/lambda/latest/dg/python-context.html
    """
    # get params REFRESH_TOKEN passed by FE in event body
    try:
        body = json.loads(event["body"])
        username = body["username"]
        password = body["password"]
        confirm_code = body["confirm_code"]
    except Exception as exc:
        raise Exception(MessageUnmarshalInputJson) from exc

    try:
        DeleteConfirmCode({"region": REGION, "user": username, "code": confirm_code})
    except Exception as e:
        raise Exception(e)

    if not re.match(PASSWORD_REGEX, password):
        AddInsertConfirmCode(info={"user": username, "confirm_code": confirm_code})
        raise Exception(MessageInvalidPassword)

    try:
        response = cog_provider_client.admin_set_user_password(
            UserPoolId=USERPOOLID, Username=username, Password=password, Permanent=True
        )
        print(response)
    except Exception as exc:
        print(exc)
        AddInsertConfirmCode(info={"user": username, "confirm_code": confirm_code})
        raise Exception(MessageForgotPasswordConfirmcodeFailed) from exc

    return generate_response(
        message=MessageForgotPasswordConfirmcodeSuccessfully, headers=RESPONSE_HEADER
    )
